src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -56,211 +88,791 @@ protected Marshaller()
+                    if (elementMarshallerKind == MarshallerKind.InvalidUnsafeType)
+                        return MarshallerKind.InvalidUnsafeType;
+
+                    if (elementMarshallerKind == MarshallerKind.AnsiChar)
+                        return MarshallerKind.AnsiCharArray;
+                    else if (elementMarshallerKind == MarshallerKind.UnicodeChar)
+                        // Arrays of unicode char should be marshalled as blittable arrays
                         return MarshallerKind.BlittableArray;
+                    else
+                        return MarshallerKind.Array;
+                    }
+
+                    case NativeType.ByValArray:         // fix sized array
+                    {
+                        var arrayType = (ArrayType)type;
+                        elementMarshallerKind = GetArrayElementMarshallerKind(
+                            arrayType,
+                            arrayType.ElementType,
+                            marshalAs,
+                            methodData,
+                            isField);
+
+                        // If element is invalid type, the array itself is invalid
+                        if (elementMarshallerKind == MarshallerKind.InvalidUnsafeType)
+                            return MarshallerKind.InvalidUnsafeType;
+
+                        if (elementMarshallerKind == MarshallerKind.AnsiChar)
+                            return MarshallerKind.ByValAnsiCharArray;
+                        else
+                            return MarshallerKind.ByValArray;
                     }
+
+                    default:
+                        return MarshallerKind.InvalidUnsafeType;
                 }
             }
 
-            if (type.IsByRef)
+            if (type.IsEnum)
+                return MarshallerKind.Enum;
+
+            if (type.IsDelegate)
             {
-                var byRefType = (ByRefType)type;
-                if (MarshalHelpers.IsBlittableType(byRefType.ParameterType))
-                    return MarshallerKind.BlittableByRef;
+                if (nativeType == NativeType.NotSet || nativeType == NativeType.Func)
+                    return MarshallerKind.FunctionPointer;
+                else
+                    return MarshallerKind.InvalidUnsafeType;
+            }
 
-                if (byRefType.ParameterType == context.GetWellKnownType(WellKnownType.Char))
+            if (type.Category == TypeFlags.Class)
+            {
+                if (methodData.IsSafeHandle(type))
                 {
-                    if (PInvokeMethodData.GetCharSet() == PInvokeAttributes.CharSetUnicode)
+                    if (nativeType == NativeType.NotSet)
+                        return MarshallerKind.SafeHandle;
+                    else
+                        return MarshallerKind.InvalidUnsafeType;
+                }
+/*
+                 TODO: Bring CriticalHandle to CoreLib
```

Comment:
It would be nice to track the todo with a Github issue and paste the issue in the comments